File: services/inventory/internal/db/migrate.go
package db

import (
	"context"
	"fmt"
	"sort"

	"inventory/migrations"
)

// Migrate applies every *.sql file embedded in the migrations package, in
// filename order, tracking applied files in a schema_migrations table so
// re-running on every startup is a no-op once a migration has landed.
//
// This is a minimal, dependency-free migration runner appropriate for a
// learning project; a production system would use a dedicated migration
// tool.
func Migrate(ctx context.Context, pool *Pool) error {
	if _, err := pool.Exec(ctx, `
		CREATE TABLE IF NOT EXISTS schema_migrations (
			filename    TEXT PRIMARY KEY,
			applied_at  TIMESTAMPTZ NOT NULL DEFAULT now()
		)
	`); err != nil {
		return fmt.Errorf("failed to ensure schema_migrations table: %w", err)
	}

	entries, err := migrations.FS.ReadDir(".")
	if err != nil {
		return fmt.Errorf("failed to read embedded migrations: %w", err)
	}

	names := make([]string, 0, len(entries))
	for _, e := range entries {
		if !e.IsDir() && len(e.Name()) > 4 && e.Name()[len(e.Name())-4:] == ".sql" {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)

	for _, name := range names {
		var alreadyApplied bool
		err := pool.QueryRow(ctx, `SELECT EXISTS (SELECT 1 FROM schema_migrations WHERE filename = $1)`, name).Scan(&alreadyApplied)
		if err != nil {
			return fmt.Errorf("failed to check migration status for %s: %w", name, err)
		}
		if alreadyApplied {
			continue
		}

		sqlBytes, err := migrations.FS.ReadFile(name)
		if err != nil {
			return fmt.Errorf("failed to read migration %s: %w", name, err)
		}

		tx, err := pool.Begin(ctx)
		if err != nil {
			return fmt.Errorf("failed to begin transaction for migration %s: %w", name, err)
		}

		if _, err := tx.Exec(ctx, string(sqlBytes)); err != nil {
			_ = tx.Rollback(ctx)
			return fmt.Errorf("failed to apply migration %s: %w", name, err)
		}

		if _, err := tx.Exec(ctx, `INSERT INTO schema_migrations (filename) VALUES ($1)`, name); err != nil {
			_ = tx.Rollback(ctx)
			return fmt.Errorf("failed to record migration %s: %w", name, err)
		}

		if err := tx.Commit(ctx); err != nil {
			return fmt.Errorf("failed to commit migration %s: %w", name, err)
		}
	}

	return nil
}

File: services/auth/src/db/migrate.ts
import { readdirSync, readFileSync } from 'fs';
import { join } from 'path';
import { Pool } from 'pg';
import { config } from '../config';
import { logger } from '../config/logger';

/**
 * Minimal, dependency-free migration runner: applies every .sql file in
 * migrations/ in filename order, tracked in a schema_migrations table so
 * re-running is a no-op. Good enough for a learning project; a real system
 * would use a proper migration tool.
 */
async function migrate(): Promise<void> {
  const pool = new Pool({ connectionString: config.databaseUrl });
  const migrationsDir = join(__dirname, '..', '..', 'migrations');

  try {
    await pool.query(`
      CREATE TABLE IF NOT EXISTS schema_migrations (
        filename    TEXT PRIMARY KEY,
        applied_at  TIMESTAMPTZ NOT NULL DEFAULT now()
      )
    `);

    const files = readdirSync(migrationsDir)
      .filter((f) => f.endsWith('.sql'))
      .sort();

    for (const file of files) {
      const { rows } = await pool.query('SELECT 1 FROM schema_migrations WHERE filename = $1', [file]);
      if (rows.length > 0) {
        logger.info({ file }, 'migration already applied, skipping');
        continue;
      }

      const sql = readFileSync(join(migrationsDir, file), 'utf8');
      logger.info({ file }, 'applying migration');
      await pool.query('BEGIN');
      try {
        await pool.query(sql);
        await pool.query('INSERT INTO schema_migrations (filename) VALUES ($1)', [file]);
        await pool.query('COMMIT');
      } catch (err) {
        await pool.query('ROLLBACK');
        throw err;
      }
    }

    logger.info('all migrations applied');
  } finally {
    await pool.end();
  }
}

migrate().catch((err) => {
  logger.error({ err }, 'migration failed');
  process.exit(1);
});
